### PR TITLE
Property now extends Data, remove BaseType

### PR DESCRIPTION
**Pull request by jackkoenig**
Currently a draft PR because there is more to do:
- [x] Merge some of the code in `Property` back into `Data`
- [x] Rip out `BaseType`
- [x] Make Bundles and Vecs containing Properties work
- [x] Figure out what to do with connections containing Properties

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- API modification

A new entry in the `Data` subhierarchy is a pretty substantial change that will affect users.

#### Desired Merge Strategy

- Squash

#### Release Notes

Properties can be fields of Bundles and used in wires.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/stage/phases/Elaborate.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ class Elaborate extends Phase {
          */
         case scala.util.control.NonFatal(a) =>
           if (!chiselOptions.printFullStackTrace) {
-            a.trimStackTraceToUserCode()
+            // a.trimStackTraceToUserCode()
```
Comment (jackkoenig) [inline src/main/scala/chisel3/stage/phases/Elaborate.scala:47]: TODO back this out before merging, useful for debugging 😅

File: core/src/main/scala/chisel3/IO.scala [hunk 2]
```diff
@@ -32,8 +33,8 @@ object IO {
     module match {
       case _: Class => {
         data match {
-          case _: Data => throwException(s"Class ports must be Property type, but found $data")
-          case _ => ()
+          case _: Property[_] => ()
+          case _ => throwException(s"Class ports must be Property type, but found $data")
```
Comment (azidar) [inline core/src/main/scala/chisel3/IO.scala:37]: Since we are here, can we throw an non-blocking error?

File: core/src/main/scala/chisel3/internal/MonoConnect.scala [hunk 3]
```diff
@@ -6,9 +6,9 @@ import chisel3._
 import chisel3.experimental.{Analog, BaseModule, SourceInfo}
 import chisel3.internal.containsProbe
 import chisel3.internal.Builder.pushCommand
-import chisel3.internal.firrtl.{Connect, Converter, DefInvalid}
+import chisel3.internal.firrtl.{Connect, Converter, DefInvalid, PropAssign}
 import chisel3.experimental.dataview.{isView, reify, reifyToAggregate}
-import chisel3.properties.Property
+import chisel3.properties.{Class, Property}
```
Comment (azidar) [inline core/src/main/scala/chisel3/internal/MonoConnect.scala:11]: I don't know if this is the place to nit, but `Class` is a perhaps undesirable name, even if it is descriptive, just for its meaning as Reflection of a java Class.

File: core/src/main/scala/chisel3/internal/MonoConnect.scala [hunk 4]
```diff
@@ -99,7 +99,7 @@ private[chisel3] object MonoConnect {
     sourceInfo:  SourceInfo,
     sink:        Data,
     source:      Data,
-    context_mod: RawModule
+    context_mod: BaseModule
```
Comment (azidar) [inline core/src/main/scala/chisel3/internal/MonoConnect.scala:102]: What's the purpose for this change? I'm fine with it, just curious.

File: core/src/main/scala/chisel3/internal/firrtl/IR.scala [hunk 5]
```diff
@@ -175,7 +175,7 @@ case class SLit(n: BigInt, w: Width) extends LitArg(n, w) {
   * These are not LitArgs, because not all property literals are integers.
   */
 private[chisel3] case class PropertyLit[T, U](
-  propertyType: PropertyTypeclass[_] { type Underlying = U; type Type = T },
+  propertyType: PropertyTypeclass[T] { type Underlying = U },
```
Comment (azidar) [inline core/src/main/scala/chisel3/internal/firrtl/IR.scala:178]: What is this change for? I'm confused by it.

File: src/test/scala/chiselTests/properties/PropertySpec.scala [hunk 6]
```diff
@@ -213,13 +213,14 @@ class PropertySpec extends ChiselFlatSpec with MatchesAndOmits {
   }
 
   it should "fail to compile when connecting Property types of different types" in {
```
Comment (azidar) [inline src/test/scala/chiselTests/properties/PropertySpec.scala:215]: Fix test name, in that it should pass?

File: core/src/main/scala/chisel3/IO.scala (hunk 2 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/IO.scala:37]: Non-blocking meaning recoverable so that we can aggregate more errors? Sure.

File: core/src/main/scala/chisel3/internal/MonoConnect.scala (hunk 3 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/MonoConnect.scala:11]: It's worth thinking about but it's a separate discussion from this PR. Lets discuss changing it offline.

File: core/src/main/scala/chisel3/internal/MonoConnect.scala (hunk 4 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/MonoConnect.scala:102]: Class extends BaseModule and connections between Properties are legal within Classes.

File: core/src/main/scala/chisel3/properties/Property.scala [hunk 7]
```diff
@@ -307,24 +263,19 @@ abstract class Property[T] extends BaseType { self =>
...
-      val value = valueOpt
-    }
-  }
+  private[chisel3] def makeWithValueOpt[T](valueOpt: Option[T])(implicit _tpe: PropertyType[T]): Property[T] =
+    new Property[T](_tpe, valueOpt)
 
   /** Create a new Property based on the type T.
     */
-  def apply[T]()(implicit tpe: PropertyType[T]): Property[tpe.Type] = {
+  def apply[T]()(implicit tpe: PropertyType[T]): Property[T] = {
```
Comment (albertchen-sifive) [inline core/src/main/scala/chisel3/properties/Property.scala:271]: The path dependent type was necessary before so that e.g. `IO(Output(Property[Int])) := Property(Property[Int]())` would work since `:=` only worked for `Property[T]` with the same type parameter. Now that `:=` works to connect any `Property[_]` regardless of type parametrization, it is not necessary for the return types to be dependent.

But the return type is still important for users that want to layer extra type safety on top of this API e.g. 
```scala
def connectProp(prop: Property[Seq[Int]]) =
  io := prop

// this would not type check anymore
connectProp(Property(Seq(Property[Int](123), IO(Input(Property[Int]())))
```

So I think we still want to keep the return type as `Property[tpe.Type]`, or just get rid of the type parameter entirely `Property[T]` -> `Property` since the core operations on `Property` aren't parameterized on `T` anyway.

Comment (jackkoenig) [reply core/src/main/scala/chisel3/properties/Property.scala:271]: Oh I understand now. Yeah even though `:=` works for any `Property` (because it works for any `Data`), the connectable operators are still typesafe, so I can test for this behavior by using `:#=` instead of `:=` connecting a `Property[Seq[Property[A]]]` to a `Property[Seq[A]]`.

Thanks for the explanation! I will restore the path dependent type.

File: core/src/main/scala/chisel3/internal/firrtl/IR.scala (hunk 5 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:178]: I originally removed the Path dependent type of Property because it didn't seem necessary and this was part of that. However, [Albert pointed out](https://github.com/chipsalliance/chisel/pull/3514/files#r1311152427) why it is, in fact, needed so I'm restoring it and backing out this change.

File: src/test/scala/chiselTests/properties/PropertySpec.scala (hunk 6 above)
Comment (jackkoenig) [reply src/test/scala/chiselTests/properties/PropertySpec.scala:215]: It actually fails when using connectable operators so I'm just turning the `:=` into a `:#=` and restoring the failing test.

The reason for the change is that before `Property` was `Data`, it had it's own `:=` which was type safe. Now that it's a `Data`, `:=` is not type safe, but connectable operators are so we have to use them to illustrate that behavior.

**Comment by jackkoenig** [comment]
Note that I carefully reconstructed the order of fields in `Data` from before `BaseType` was added (and took some of the fields out in https://github.com/chipsalliance/chisel/pull/3429), you can diff `Data.scala` with the prior commit (https://github.com/chipsalliance/chisel/commit/878733f86cb4582f7309de35384af64ed6bb16b7) to see.

File: core/src/main/scala/chisel3/internal/BiConnect.scala [hunk 8]
```diff
@@ -10,6 +10,7 @@ import chisel3.internal.firrtl.{Connect, Converter, DefInvalid}
 
 import scala.language.experimental.macros
 import _root_.firrtl.passes.CheckTypes
+import chisel3.properties.Property
```
Comment (seldridge) [inline core/src/main/scala/chisel3/internal/BiConnect.scala:13]: Mega-nit: group with other `chisel3` imports.

File: src/test/scala/chiselTests/properties/PropertySpec.scala [hunk 9]
```diff
@@ -338,11 +337,11 @@ class PropertySpec extends ChiselFlatSpec with MatchesAndOmits {
...
       "output b : List<Map<List<Integer>>>",
       "output c : List<Map<List<Integer>>>",
-      "output d : List<Map<List<Integer>>>",
+      // "output d : List<Map<List<Integer>>>",
       """propassign a, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))""",
       """propassign b, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))""",
-      """propassign c, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))""",
-      """propassign d, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))"""
+      """propassign c, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))"""
+      // """propassign d, List<Map<List<Integer>>>(Map<List<Integer>>("foo" -> List<Integer>(Integer(123))))"""
```
Comment (seldridge) [inline src/test/scala/chiselTests/properties/PropertySpec.scala:344]: Nit: remove comments or do something different here.

**Review by seldridge — APPROVED** [review]
LGTM

Do I read this correctly in that this is working for `Wire`, too?

File: core/src/main/scala/chisel3/internal/BiConnect.scala (hunk 8 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/BiConnect.scala:13]: That's what I get for using VScode 😅

File: src/test/scala/chiselTests/properties/PropertySpec.scala (hunk 9 above)
Comment (jackkoenig) [reply src/test/scala/chiselTests/properties/PropertySpec.scala:344]: Lol I meant to uncomment these but forgot, will fix

**Comment by jackkoenig** [comment]
> Do I read this correctly in that this is working for Wire, too?

You do indeed, I'll mention that in the release notes.

**Review by azidar — APPROVED** [review]
